### PR TITLE
Verify output artifacts as found from cache before making execution decision

### PR DESCRIPTION
**Pull request by micahjsmith**
If cache is enabled, and if matching artifacts are found in the cache for a component's outputs, verify the output artifacts before making an execution decision. This avoids a situation in which the artifacts are missing (no object exists at the URI associated with the artifact due to a deletion process outside of the pipeline) and the downstream component that uses this component's outputs fails.

* since the downstream consumer of the cached artifacts will use `BaseDriver.verify_input_artifacts` to verify the artifacts, use the same method when checking the cache to "fail fast" (avoid using the cache)
* note that MLMD ArtifactState is not used at all in the BaseDriver logic. perhaps in a follow-up PR, we can add (1) checks when getting artifacts from cache that the objects are marked LIVE/PUBLISHED and (2) for artifacts from cache that are deemed to be missing given the logic in the current PR, update their state in MLMD to MISSING

cc @1025KB to follow up on our discussion on Wednesday

### CODE REVIEW

File: tfx/dsl/components/base/base_driver.py [hunk 1]
```diff
@@ -292,7 +292,17 @@ def pre_execution(
...
-    if output_artifacts is not None:
+
+      # Check that cached output artifacts will actually be considered a cache hit by downstream
+      # components
+      if output_artifacts is not None:
+        try:
+          self.verify_input_artifacts(artifacts_dict=output_artifacts)
+          use_cached_results = True
+        except RuntimeError:
+          use_cached_results = False
```
Comment (1025KB) [inline tfx/dsl/components/base/base_driver.py:303]: let's log a info message here for debugging reason

**Review by 1025KB — COMMENTED** [review]
Thanks!!

**Review by 1025KB — COMMENTED** [review]
Note that [here](https://github.com/tensorflow/tfx/blob/master/tfx/orchestration/portable/cache_utils.py#L173) needs similar updates too.

base_driver is the old path, new orchestrator like LocalDagRunner uses above portable code

**Comment by micahjsmith** [comment]
> Note that [here](https://github.com/tensorflow/tfx/blob/master/tfx/orchestration/portable/cache_utils.py#L173) needs similar updates too.
> 
> base_driver is the old path, new orchestrator like LocalDagRunner uses above portable code

thanks for the review, will make those changes in a bit!

**Comment by micahjsmith** [comment]
@1025KB I'm looking at the differences between base_driver and cache_utils. I was hoping that base_driver could be refactored to use cache_utils (in the same way it is used by tfx.orchestration.portable.launcher), but it seems there are some significant differences, such as the way they compute cache keys from context and the steps in get_cached_context. unless you tell me otherwise (i.e. that I should look further into refactoring them together), I'm just going to copy/paste my new logic into cache_utils

**Comment by 1025KB** [comment]
yep, it's different code path at this point, so logic duplication sounds good. Thanks!!

**Comment by micahjsmith** [comment]
@1025KB ready for re-review!

* refactored the logic of checking whether artifacts exist into `artifact_utils.verify_artifacts`
* use `artifact_utils.verify_artifacts` in three locations: `BaseDriver.verify_input_artifacts`, `BaseDriver.pre_execution`, and `cache_utils.get_cached_outputs`
* add tests
* fix all extant lint issues in touched files, sorry for the large diff as a result!

note that the logic differs between `cache_utils.get_cached_outputs` and `BaseDriver.pre_execution`/`metadata.get_cached_outputs`. in the former case, all the matching executions are checked for valid artifacts. in the latter case, only the most recent matching execution is checked for valid artifacts.

File: tfx/orchestration/portable/cache_utils.py [hunk 2]
```diff
@@ -166,11 +166,15 @@ def get_cached_outputs(
...
   # Defensively traverses candidate executions and returns once we find an
-  # execution with valid outputs.
+  # execution with valid outputs that can be confirmed to still exist.
   for execution in cached_executions:
     cached_output_artifacts = _get_outputs_of_execution(metadata_handler,
                                                         execution.id)
     if cached_output_artifacts is not None:
-      return cached_output_artifacts
+      try:
+        artifact_utils.verify_artifacts(cached_output_artifacts)
```
[the remark this answers is not in the document]
Comment (micahjsmith) [reply tfx/orchestration/portable/cache_utils.py:175]: added!

File: tfx/orchestration/portable/cache_utils_test.py [hunk 3]
```diff
@@ -24,9 +24,10 @@
 from tfx.proto.orchestration import pipeline_pb2
 from tfx.types import standard_artifacts
 from tfx.utils import test_case_utils
+
 from google.protobuf import text_format
 from ml_metadata.proto import metadata_store_pb2
-
+import tensorflow as tf
```
Comment (1025KB) [inline tfx/orchestration/portable/cache_utils_test.py:30]: import tf is quite heavy, let's drop it if not necessary?

Comment (1025KB) [inline tfx/orchestration/portable/cache_utils_test.py:30]: keep previous order?

**Review by rcrowe-google — APPROVED** [review]
LGTM, thanks!

File: tfx/types/artifact_utils_test.py [hunk 4]
```diff
@@ -179,6 +179,21 @@ def testIsArtifactVersionOlderThan(self):
     self.assertFalse(
         artifact_utils.is_artifact_version_older_than(examples, '0.1'))
 
+  @mock.patch("tfx.types.artifact_utils.fileio", autospec=True)
```
Comment (jiyongjung0) [inline tfx/types/artifact_utils_test.py:182]: How about adding some test cases for the error case? For example uri is missing or the file doesn't exist.

**Review by jiyongjung0 — COMMENTED** [review]
Thank you for the great contribution!

File: tfx/orchestration/portable/cache_utils_test.py (hunk 3 above)
Comment (micahjsmith) [reply tfx/orchestration/portable/cache_utils_test.py]: done!

thought I understood the pattern as stdlib, firstpart, thirdparty which would place tf in thirdparty group. but it's not consistent in other files, so not my problem 🤷

File: tfx/types/artifact_utils_test.py (hunk 4 above)
Comment (micahjsmith) [reply tfx/types/artifact_utils_test.py:182]: added!

**Review by 1025KB — APPROVED** [review]

**Comment by rcrowe-google** [comment]
It seems that we always assume that artifact payloads are in file systems.  Should we?

**Comment by 1025KB** [comment]
> It seems that we always assume that artifact payloads are in file systems. Should we?

at this point it's still the case

**Pull request merged**